File: Searching and Sorting Algorithms/codigo_cs/11NextGreatElement.cs
using System;
using System.Collections.Generic;
using System.Linq;
public class Test11
{
    public static void NGE()
    {
        // your code goes here
        int n = int.Parse(Console.ReadLine());
        int[] arr = Array.ConvertAll(Console.ReadLine().Split(), int.Parse);

        int[] res = Enumerable.Repeat(-1, n).ToArray();
        Stack<int> stack = new Stack<int>();
        for (int i = n - 1; i >= 0; i--)
        {
            while (stack.Count > 0 && stack.Peek() <= arr[i])
            {
                stack.Pop();
            }
            if (stack.Count > 0)
            {
                res[i] = stack.Peek();
            }
            stack.Push(arr[i]);
        }
        Console.WriteLine(string.Join(" ", res));

    }
}

File: Searching and Sorting Algorithms/11NextGreatElement.py
n=int(input())
arr=list(map(int,input().split()))
stack=[]
res=[-1]*n
for i in range(n-1,-1,-1):
    while stack and stack[-1]<=arr[i]:
        stack.pop()
    if stack:
        res[i]=stack[-1]
    stack.append(arr[i])

print(*res)